python/ray/rllib/env/external_multi_agent_env.py

```diff
@@ -0,0 +1,262 @@
+
+    Examples:
+        >>> register_env("my_env", lambda config: YourExternalEnv(config))
+        >>> agent = DQNAgent(env="my_env")
+        >>> while True:
+              print(agent.train())
+    """
+
+    @PublicAPI
+    def __init__(self, action_space, observation_space, max_concurrent=100):
+        """Initialize an external env.
+
+        ExternalEnv subclasses must call this during their __init__.
+
+        Arguments:
+            action_space (gym.Space): Action space of the env.
+            observation_space (gym.Space): Observation space of the env.
+            max_concurrent (int): Max number of active episodes to allow at
+                once. Exceeding this limit raises an error.
+        """
+
+        threading.Thread.__init__(self)
+        self.daemon = True
+        self.action_space = action_space
+        self.observation_space = observation_space
+        self._episodes = {}
+        self._finished = set()
+        self._results_avail_condition = threading.Condition()
+        self._max_concurrent_episodes = max_concurrent
+
+        # we require to know all agents' spaces
+        if isinstance(self.action_space, dict) or isinstance(self.observation_space, dict):
+            assert(self.action_space.keys() == self.observation_space.keys())
+
+    @PublicAPI
+    def run(self):
+        """Override this to implement the run loop.
+
+        Your loop should continuously:
+            1. Call self.start_episode(episode_id)
+            2. Call self.get_action(episode_id, obs)
+                    -or-
+                    self.log_action(episode_id, obs, action)
+            3. Call self.log_returns(episode_id, reward)
+            4. Call self.end_episode(episode_id, obs)
+            5. Wait if nothing to do.
+
+        Multiple episodes may be started at the same time.
+        """
+        raise NotImplementedError
+
+    @PublicAPI
+    def start_episode(self, episode_id=None, training_enabled=True):
+        """Record the start of an episode.
+
+        Arguments:
+            episode_id (str): Unique string id for the episode or None for
+                it to be auto-assigned.
+            training_enabled (bool): Whether to use experiences for this
+                episode to improve the policy.
+
+        Returns:
+            episode_id (str): Unique string id for the episode.
+        """
+
+        if episode_id is None:
+            episode_id = uuid.uuid4().hex
+
+        if episode_id in self._finished:
+            raise ValueError(
+                "Episode {} has already completed.".format(episode_id))
+
+        if episode_id in self._episodes:
+            raise ValueError(
+                "Episode {} is already started".format(episode_id))
+
+        self._episodes[episode_id] = _ExternalEnvEpisode(
+            episode_id, self._results_avail_condition, training_enabled)
+
+        return episode_id
+
+    @PublicAPI
+    def get_action(self, episode_id, observation):
```

Comment:
One design decision here is whether you should be getting the action of multiple agents at once (as opposed to one at a time).

I think it may be easier to enforce all the observations of all agents acting for the "episode step" must be provided at once in this call. Otherwise, it becomes unclear which agent actions go in which episode timestep.

Reply:
While I think it would be easier to enforce an observation of all agents in one call to ``get_action(...)``, this would not allow the same "level" of asynchronism as offered with ``MultiAgentEnv``, right?

Reply:
Hmm I was thinking you could allow a subset to be passed in. Similar to how step() in MultiAgentEnv returns a subset of the agents in the env, get_action() could take a subset of the agents as well. So the level of asynchrony would be matching.

Reply:
That makes sense, thanks for clarification.

As the logic of ExternalMultiAgentEnv would then be analogous to MultiAgentEnv, I'd opt for that.
Currently, this is working as expected: one can pass a subset of agent observations and get that subset of agent actions back.